### PR TITLE
STM32L475 ADU sample

### PR DESCRIPTION
**Pull request by vaavva**
## Purpose

* Adds ADU sample for the STM32L475
* Readme to come in a follow-up PR, as this only merges previously reviewed code

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Review by wduraes — APPROVED** [review]
since code changes were previously approved, I'm approving and will wait for the merge so I can test on the device.

File: demos/projects/ST/b-l475e-iot01a/ADU.md
Comment (danewalton) [inline demos/projects/ST/b-l475e-iot01a/ADU.md:9]: nit: make 1.6 -> 1.1

**Review by danewalton — APPROVED** [review]
Before merging, please restore this: 

https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/d73de4bcf47e68503fa4afa43265b2ef8eb6ca93/.github/workflows/ci_tests_linux.yml#L22-L23

to add st to the gate. You can remove the line after it which had dropped st. You'll have to add `iot-middleware-sample-adu` to here as well : https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/d73de4bcf47e68503fa4afa43265b2ef8eb6ca93/.github/scripts/ci_tests.sh#L83

as a new check

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:70]: Is this tracked in any way by ADU? Should it be more specific (i.e. including version, the fact it's a sample, etc)

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:94]: Please consider increasing the timeouts, especially for WiFi communication.

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:117]: Same as above.

File: demos/projects/ST/b-l475e-iot01a/config/demo_config.h
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/config/demo_config.h:518]: Should these include some information about our SDK and version for telemetry purposes?

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c:5]: Nit: we should remove whitespace between includes.

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c:23]: Should this be defined as 2*sizeof(long) or uintptr_t ?

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (CIPop) [inline demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c:220]: This should not be Information but error (or at least warning). We should also try to log a single time instead of many log lines.

Also, formatting should only be done by AZLogInfo and not printf - a log listener may direct these messages elsewhere.

**Review by CIPop — COMMENTED** [review]
Left a few comments.

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:94]: Any suggestions on a better value?

File: demos/projects/ST/b-l475e-iot01a/config/demo_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/demo_config.h:518]: This is what shows up in Azure Portal when you select device information, so I think it's probably better to be just about the device. We could definitely open this as a topic for all of the ADU samples (currently the rest of them are similar to this)

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c:220]: This was done to match the verification in the ESP32 sample, but let me see what I can do to modify it and then make those changes there as well

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:70]: I'm not sure - @danewalton, do you know the answer to this?

Comment (danewalton) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:70]: to my knowledge, no they don't track it

Comment (CIPop) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:70]: @wduraes - FYI, in case there's something we need to track.

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (CIPop) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:94]: Maybe I misunderstood what these mean but I'd give it a try with 500/1000ms, especially since offline we were talking about some disconnects in certain cases.

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c]: I was able to change the first line to an error (I wasn't sure if the rest should be categorized more as additional info or more error lines).
I couldn't figure out a way to print the whole hash so we could use one logError for the whole bit of info - I think I'm running into this issue: https://eevblog.com/forum/microcontrollers/printf-uint64_t-with-arm-gcc/ 
> AZLogError( ( "SHAs do not match\r\nWanted: %.*x\r\nCalculated: %.*x\r\n", azureiotflashSHA_256_SIZE, *ucDecodedManifestHash, azureiotflashSHA_256_SIZE, *ucCalculatedHash ) );

I also tried to at least convert the printfs to AzureLogInfos, but that adds a newline at the end, so the hash ends up like this instead of on the same line 
> ea
fa
d1
fa
cd
ea
6a
cc
ab
33
...

If you have any suggestions on how to solve either of these issues, let me know!

File: demos/projects/ST/b-l475e-iot01a/config/demo_config.h
Comment (CIPop) [reply demos/projects/ST/b-l475e-iot01a/config/demo_config.h:518]: In that case it should definitely be something like `Azure IoT SDK FreeRTOS ADU Sample` (or similar w/o spaces) to make sure there is no confusion between prod and test devices.

File: demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c
Comment (CIPop) [reply demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c]: Ah, in that case you could consider the issue a future enhancement and track separately. 

In general, we should never log more than one line per error/info message. This helps log analysis tools that usually expect a single record per message.

**Review by CIPop — APPROVED** [review]
Thanks @vaavva !

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:94]: The issues I was running into were because my wifi hotspot is only connectable when my phone is open to the hotspot page for some reason 🙃 And the second issue was with disconnects during download, but that issue I think should be fixed with retries, not a longer timeout. So I'm not quite sure how to test whether a larger number is sufficient or not (especially since the comments for this say to keep it small)
@danewalton, did you have a way to repro the issues you were seeing?

File: demos/projects/ST/b-l475e-iot01a/config/demo_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/demo_config.h:518]: I'll open a work item for this and the printing issue to solve these issues in a consistent way across the samples

File: demos/projects/ST/b-l475e-iot01a/config/core_http_config.h
Comment (vaavva) [reply demos/projects/ST/b-l475e-iot01a/config/core_http_config.h:94]: Discussed offline - will leave as 10U for now

**Pull request merged**
